api/v1beta1/azuremachine_default.go

```diff
@@ -79,10 +88,32 @@ func (s *AzureMachineSpec) SetDataDisksDefaults() {
 }
 
 // SetIdentityDefaults sets the defaults for VM Identity.
-func (s *AzureMachineSpec) SetIdentityDefaults() {
+func (s *AzureMachineSpec) SetIdentityDefaults(subscriptionID string) {
+	// Ensure the deprecated fields and new fields are not populated simultaneously
+	if s.RoleAssignmentName != "" && s.SystemAssignedIdentityRole != nil && s.SystemAssignedIdentityRole.Name != "" {
+		// Both the deprecated and the new fields are both set, return without changes
+		// and reject the request in the validating webhook which runs later.
+		return
+	}
 	if s.Identity == VMIdentitySystemAssigned {
-		if s.RoleAssignmentName == "" {
-			s.RoleAssignmentName = string(uuid.NewUUID())
+		if s.SystemAssignedIdentityRole == nil {
+			s.SystemAssignedIdentityRole = &SystemAssignedIdentityRole{}
+		}
+		if s.RoleAssignmentName != "" {
+			// Move the existing value from the deprecated RoleAssignmentName field.
+			s.SystemAssignedIdentityRole.Name = s.RoleAssignmentName
+			s.RoleAssignmentName = ""
+		} else if s.SystemAssignedIdentityRole.Name == "" {
+			// Default role name to the role assignment name.
```

Comment:
I'm confused by this comment, because I guess it refers to an implied action: leaving `s.RoleAssignmentName` alone. But the next line is assigning a new UUID to a similar field... Maybe we can add something to the comment like `and assign a UUID`?

Or maybe I'm misunderstanding entirely. :-)

Reply:
This comment is definitely misleading. Since we're deprecating `s.RoleAssignmentName`, we are copying the name over to the new `s.SystemAssignedIdentityRole` field. Otherwise if the name field is empty, we want to assign a UUID. I'll go ahead and fix the comment!